erpnext/stock/report/available_serial_no/available_serial_no.py

```diff
@@ -0,0 +1,312 @@
+from erpnext.stock.doctype.serial_no.serial_no import (
+	get_serial_nos,
+	get_serial_nos_from_serial_and_batch_bundle,
+)
+from erpnext.stock.doctype.stock_reconciliation.stock_reconciliation import get_stock_balance_for
+from erpnext.stock.report.stock_ledger.stock_ledger import (
+	check_inventory_dimension_filters_applied,
+	get_item_details,
+	get_item_group_condition,
+	get_opening_balance,
+	get_opening_balance_from_batch,
+	get_stock_ledger_entries,
+)
+from erpnext.stock.utils import (
+	is_reposting_item_valuation_in_progress,
+	update_included_uom_in_report,
+)
+
+
+def execute(filters=None):
+	is_reposting_item_valuation_in_progress()
+	include_uom = filters.get("include_uom")
+	columns = get_columns(filters)
+	items = get_items(filters)
+	sl_entries = get_stock_ledger_entries(filters, items)
+	item_details = get_item_details(items, sl_entries, include_uom)
+	if filters.get("batch_no"):
+		opening_row = get_opening_balance_from_batch(filters, columns, sl_entries)
+	else:
+		opening_row = get_opening_balance(filters, columns, sl_entries)
+
+	precision = cint(frappe.db.get_single_value("System Settings", "float_precision"))
+
+	data = []
+	conversion_factors = []
+	if opening_row:
+		data.append(opening_row)
+		conversion_factors.append(0)
+
+	actual_qty = stock_value = 0
+	if opening_row:
+		actual_qty = opening_row.get("qty_after_transaction")
+		stock_value = opening_row.get("stock_value")
+
+	available_serial_nos = {}
+	inventory_dimension_filters_applied = check_inventory_dimension_filters_applied(filters)
+
+	batch_balance_dict = frappe._dict({})
+	if actual_qty and filters.get("batch_no"):
+		batch_balance_dict[filters.batch_no] = [actual_qty, stock_value]
+
+	for sle in sl_entries:
+		item_detail = item_details[sle.item_code]
+
+		sle.update(item_detail)
+
+		if filters.get("batch_no") or inventory_dimension_filters_applied:
+			actual_qty += flt(sle.actual_qty, precision)
+			stock_value += sle.stock_value_difference
+			if sle.batch_no:
+				if not batch_balance_dict.get(sle.batch_no):
+					batch_balance_dict[sle.batch_no] = [0, 0]
+
+				batch_balance_dict[sle.batch_no][0] += sle.actual_qty
+
+			if sle.voucher_type == "Stock Reconciliation" and not sle.actual_qty:
+				actual_qty = sle.qty_after_transaction
+				stock_value = sle.stock_value
+
+			sle.update({"qty_after_transaction": actual_qty, "stock_value": stock_value})
+
+		sle.update({"in_qty": max(sle.actual_qty, 0), "out_qty": min(sle.actual_qty, 0)})
+
+		update_available_serial_nos(available_serial_nos, sle)
+
+		if sle.actual_qty:
+			sle["in_out_rate"] = flt(sle.stock_value_difference / sle.actual_qty, precision)
+
+		elif sle.voucher_type == "Stock Reconciliation":
+			sle["in_out_rate"] = sle.valuation_rate
+
+		data.append(sle)
+
+		if include_uom:
+			conversion_factors.append(item_detail.conversion_factor)
+
+	update_included_uom_in_report(columns, data, include_uom, conversion_factors)
+	return columns, data
```

Comment:
Number of lines in the execute function is 97 which is huge, split it into small functions